File: nodejs-project/crawlerIndex.js
"use strict";

/**
 * Crawler that collects all article URLs from shacho.osakazine.net album pages.
 * Converted from crawler_index.py (Scrapy -> axios + cheerio, since Scrapy
 * has no direct Node.js equivalent).
 */

const fs = require("fs");
const path = require("path");
const axios = require("axios");
const cheerio = require("cheerio");
const logUtils = require("./logUtils");

const BASE_URL = "https://shacho.osakazine.net";
const START_URL = `${BASE_URL}/album.html`;
// Lambda's filesystem is read-only except for /tmp
const OUTPUT_FILE = path.join("/data", "article_urls.json");
// const OUTPUT_FILE = path.join('/tmp', 'data', 'article_urls.json');
const USER_AGENT =
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) " +
  "AppleWebKit/537.36 (KHTML, like Gecko) " +
  "Chrome/124.0.0.0 Safari/537.36";

const DATE_PATTERN = /(\d{4}\/\d{2}\/\d{2})/;
const DOWNLOAD_DELAY_MS = 1000; // mirrors Scrapy's DOWNLOAD_DELAY

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** "2026/04/10" -> "2026-04-10", validating the date is real. */
function formatDate(title) {
  const m = DATE_PATTERN.exec(title);
  if (!m) return "";
  const [y, mo, d] = m[1].split("/").map(Number);
  const dt = new Date(Date.UTC(y, mo - 1, d));
  const valid =
    dt.getUTCFullYear() === y &&
    dt.getUTCMonth() === mo - 1 &&
    dt.getUTCDate() === d;
  if (!valid) return "";
  return `${String(y).padStart(4, "0")}-${String(mo).padStart(2, "0")}-${String(d).padStart(2, "0")}`;
}

function resolveUrl(base, href) {
  try {
    return new URL(href, base).toString();
  } catch (e) {
    return href;
  }
}

async function fetchPage(url) {
  return axios.get(url, {
    headers: { "User-Agent": USER_AGENT },
    timeout: 30000,
    validateStatus: () => true, // we check status.404 ourselves, like handle_httpstatus_list
  });
}

function parseAlbumPage($) {
  const articles = [];

  $("div.album div.album_frame").each((_, frame) => {
    const aTag = $(frame).find("div.album_image a").first();
    if (!aTag.length) return;

    const href = aTag.attr("href") || "";
    if (!href) return;

    const title = aTag.attr("title") || "";
    const date = formatDate(title);
    const url = href.startsWith("http") ? href : `${BASE_URL}/${href}`;

    articles.push({ url, date });
  });

  let nextHref = null;
  $("div.page_nav a").each((_, a) => {
    if (($(a).text() || "").includes("次へ")) {
      nextHref = $(a).attr("href") || "";
      return false; // break out of .each
    }
  });

  return { articles, nextHref };
}

/**
 * Walk album pages starting from START_URL, following "next page" links,
 * and save all discovered article URLs to outputFile.
 */
async function crawlIndex(outputFile = OUTPUT_FILE) {
  logUtils.crawler(`crawl_index: starting, output → ${outputFile}`);

  const allArticles = [];
  let currentUrl = START_URL;

  while (currentUrl) {
    const response = await fetchPage(currentUrl);

    if (response.status === 404) {
      logUtils.crawler("404 — stopping");
      break;
    }

    const $ = cheerio.load(response.data);
    const { articles, nextHref } = parseAlbumPage($);

    allArticles.push(...articles);
    logUtils.crawler(
      `${currentUrl}: ${articles.length} article(s) found (total ${allArticles.length})`,
    );

    if (nextHref) {
      logUtils.crawler(`following next page: ${nextHref}`);
      currentUrl = resolveUrl(currentUrl, nextHref);
      await sleep(DOWNLOAD_DELAY_MS);
    } else {
      logUtils.crawler("no next page — done");
      currentUrl = null;
    }
  }

  const records = allArticles
    .map((a) => ({
      url: a.url,
      date: a.date,
      status: "not_crawled",
      wp_post_id: null,
      wp_post_slug: null,
    }))
    .sort((a, b) => (a.date < b.date ? -1 : a.date > b.date ? 1 : 0));

  fs.mkdirSync(path.dirname(outputFile), { recursive: true });
  fs.writeFileSync(outputFile, JSON.stringify(records, null, 2), "utf-8");

  logUtils.crawler(
    `crawl_index: done — saved ${allArticles.length} URLs to ${outputFile}`,
  );

  return allArticles.map((a) => a.url);
}

module.exports = { crawlIndex, OUTPUT_FILE, START_URL, BASE_URL };

if (require.main === module) {
  crawlIndex().catch((e) => {
    logUtils.error(e.message);
    process.exitCode = 1;
  });
}

File: python-project/crawler-system/lambda_index/crawler_index.py
"""Crawler that collects all article URLs from shacho.osakazine.net album pages."""

import json
import os
import re
from datetime import datetime

import requests
from bs4 import BeautifulSoup

from utils import log_utils

BASE_URL = "https://shacho.osakazine.net"
START_URL = f"{BASE_URL}/album.html"
OUTPUT_FILE = "data/article_urls.json"

USER_AGENT = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
    "AppleWebKit/537.36 (KHTML, like Gecko) "
    "Chrome/124.0.0.0 Safari/537.36"
)

DATE_PATTERN = re.compile(r"(\d{4}/\d{2}/\d{2})")


def crawl_index(output_file: str = OUTPUT_FILE) -> list[str]:
    """Collect all article URLs from album pages."""

    if os.environ.get("AWS_EXECUTION_ENV"):
        output_file = "/tmp/article_urls.json"

    log_utils.crawler(f"crawl_index: starting, output → {output_file}")

    session = requests.Session()
    session.headers.update({"User-Agent": USER_AGENT})

    url = START_URL
    articles: list[dict] = []

    while url:
        log_utils.crawler(f"Crawling: {url}")

        try:
            response = session.get(url, timeout=30)
            response.raise_for_status()
        except requests.RequestException as e:
            log_utils.error(f"Failed to fetch {url}: {e}")
            break

        soup = BeautifulSoup(response.text, "lxml")

        count_before = len(articles)

        for frame in soup.select("div.album div.album_frame"):
            a_tag = frame.select_one("div.album_image a")

            if not a_tag:
                continue

            href = a_tag.get("href", "")
            if not href:
                continue

            title = a_tag.get("title", "")

            date_str = ""
            m = DATE_PATTERN.search(title)

            if m:
                try:
                    date_str = datetime.strptime(
                        m.group(1),
                        "%Y/%m/%d",
                    ).strftime("%Y-%m-%d")
                except ValueError:
                    pass

            article_url = href if href.startswith("http") else f"{BASE_URL}/{href}"

            articles.append(
                {
                    "url": article_url,
                    "date": date_str,
                }
            )

        found = len(articles) - count_before

        log_utils.crawler(f"{url}: {found} article(s) found (total {len(articles)})")

        next_url = None

        for a in soup.select("div.page_nav a"):
            if "次へ" in a.get_text(strip=True):
                href = a.get("href", "")
                if href:
                    next_url = href if href.startswith("http") else f"{BASE_URL}/{href}"
                break

        if next_url:
            log_utils.crawler(f"following next page: {next_url}")
        else:
            log_utils.crawler("no next page — done")

        url = next_url

    records = sorted(
        [
            {
                "url": a["url"],
                "date": a["date"],
                "status": "not_crawled",
                "wp_post_id": None,
                "wp_post_slug": None,
            }
            for a in articles
        ],
        key=lambda r: r["date"],
    )

    os.makedirs(os.path.dirname(output_file), exist_ok=True)

    with open(output_file, "w", encoding="utf-8") as f:
        json.dump(records, f, ensure_ascii=False, indent=2)

    log_utils.crawler(f"crawl_index: done — saved {len(records)} URLs to {output_file}")

    return [r["url"] for r in records]


if __name__ == "__main__":
    crawl_index()
